src/Mvc/Mvc.ViewFeatures/src/Infrastructure/HttpUriHelper.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Components.Services;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures
+{
+    internal class HttpUriHelper : UriHelperBase
+    {
+        private HttpContext _context;
+
+        public HttpUriHelper()
+        {
+        }
+
+        public void InitializeState(HttpContext context)
+        {
+            _context = context;
+            InitializeState();
+        }
+
+        protected override void InitializeState()
+        {
+            if (_context == null)
+            {
+                throw new InvalidOperationException($"'{typeof(HttpUriHelper)}' not initialized.");
+            }
+            SetAbsoluteBaseUri(GetContextBaseUri());
+            SetAbsoluteUri(GetFullUri());
+        }
+
+        private string GetFullUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(
+                request.Scheme,
+                request.Host,
+                request.PathBase,
+                request.Path,
+                request.QueryString);
+        }
+
+        private string GetContextBaseUri()
+        {
+            var request = _context.Request;
+            return UriHelper.BuildAbsolute(request.Scheme, request.Host, request.PathBase);
+        }
+
+        protected override void NavigateToCore(string uri, bool forceLoad)
+        {
+            // For now throw.
+            // We don't have a good way to do redirects on a prerrendering context.
+            // Currently there's no good way to stop the rendering process. We can fix
+            // this by passing in a cancellation token to RenderRootComponentAsync and
+            // having that checked before we await on every ProcessAsynchronousWork iteration
+            // if we choose to go that way. Upon redirect we could trigger the cancellation here
+            // and do _context.Response.Redirect(uri)
+            // The bigger (not easily solvable problem) is that the response might have already
+            // started (for example when the page/view has already been flushed) which will mean
+            // that the behavior of this feature is not reliable, and not easily to diagnose as
+            // the developer only notices it when this happens at runtime; something we don't really want.
```

Comment:
> The bigger (not easily solvable problem) is that the response might have already started

Is that really a problem, though? The prerendering system never starts emitting any response until the initial async render is completed. So the only way that the response has started is if it's for an unrelated reason, so it's no different from someone trying to do a redirection in MVC after they've already started flushing out output - the rules for that are well understood. As such it doesn't look to me like this would be unreliable at all.

If I'm understanding correctly, could we either remove or rephrase the comment? It could be misleading otherwise.

Reply:
I'll remove it for now and file an issue once we tackle routing

Reply:
> Is that really a problem, though? The prerendering system never starts emitting any response until the initial async render is completed. So the only way that the response has started is if it's for an unrelated reason, so it's no different from someone trying to do a redirection in MVC after they've already started flushing out output - the rules for that are well understood. As such it doesn't look to me like this would be unreliable at all.

Yeah, but this needs to work well with MVC and I'm pretty sure we've discouraged people in the past from doing a redirect within a View. Even if you succeed in the redirect, a lot of things can go wrong in the server and it's just walking into unknown territory.

It's not simply about performing the redirect, but about what happens on the server. How do you stop rendering, do you throw? If so you start polluting your logs with exceptions that are surreptitiously used to do flow control.

IMO we need to think about this in detail, so for now I want to just throw until we tackle it in a holistic way.